extensions/amp-video/1.0/video-wrapper.js

```diff
@@ -0,0 +1,232 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {ContainWrapper} from '../../../src/preact/component';
+import {Deferred} from '../../../src/utils/promise';
+import {MIN_VISIBILITY_RATIO_FOR_AUTOPLAY} from '../../../src/video-interface';
+import {cssText as autoplayCss} from '../../../build/video-autoplay.css';
+import {fillContentOverlay} from './video-wrapper.css';
+import {
+  parseFavicon,
+  parseOgImage,
+  parseSchemaImage,
+  setMediaSession,
+} from '../../../src/mediasession-helper';
+import {useEffect, useMemo, useRef, useState} from '../../../src/preact';
+import {useMountEffect, useResourcesNotify} from '../../../src/preact/utils';
+
+/**
+ * @param {{getMetadata: (function():!Object|undefined)}} player
+ * @param {!VideoWrapperProps} props
+ * @return {!Object};
+ */
+function getMetadata(player, props) {
+  const metadata = player.getMetadata
+    ? player.getMetadata()
+    : {
+        title: props.title || '',
+        artist: props.artist || '',
+        album: props.album || '',
+        artwork: [{src: props.artwork || props.poster || ''}],
+      };
+
+  metadata.title = metadata.title || props['aria-label'] || document.title;
+
+  metadata.artwork =
+    metadata.artwork ||
+    parseSchemaImage(document) ||
+    parseOgImage(document) ||
+    parseFavicon(document);
+
+  return metadata;
+}
+
+/**
+ * @param {!VideoWrapperProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function VideoWrapper({
+  component: Component = 'video',
+  autoplay = false,
+  controls = false,
+  noaudio = false,
+  mediasession = true,
+  'class': className,
+  style,
+  children,
+  ...rest
+}) {
+  useResourcesNotify();
+
+  const [muted, setMuted] = useState(autoplay);
+  const [metadata, setMetadata] = useState(null);
+  const [playing, setPlaying] = useState(false);
+  const [userInteracted, setUserInteracted] = useState(false);
+
+  const wrapperRef = useRef(null);
+  const playerRef = useRef(null);
+
+  const readyDeferred = useMemo(() => new Deferred(), []);
+
+  const play = useMemo(
+    () => () => {
+      readyDeferred.promise.then(() => {
+        playerRef.current.play();
+      });
+    },
+    [readyDeferred]
+  );
+
+  const pause = useMemo(
+    () => () => {
+      readyDeferred.promise.then(() => {
+        playerRef.current.pause();
+      });
+    },
+    [readyDeferred]
+  );
+
+  useEffect(() => {
+    if (mediasession && metadata && playing) {
+      setMediaSession(window, metadata, play, pause);
+    }
+  }, [mediasession, playing, metadata, play, pause]);
+
+  return (
+    <ContainWrapper
+      contentRef={wrapperRef}
+      class={className}
+      style={style}
```

Comment:
If you're doing `style` separately, we should also do `className`. But I'm actually thinking, why not just do `{...rest}` twice? It has a risk of repeating `aria-label` and `data-some` twice, but would that be a big issue?

Reply:
This is interesting, made the change.

- `aria-label` can't repeat as it'd be spoken back twice.
- `style`, `class` and `data-*` *could* be ok as long as we don't reasonably expect style/CSS selectors that apply to both to have side-effects from duplicating on a child/parent, e.g. same `border-*` on both elements.

Reply:
No. Sorry about that. It was a bad suggestion. We definitely can't duplicate `style` and `className`. So they should definitely be passed in explicitly. So, I'd say, this case is unlike other wrappers and should just take `style`, `className` and send everything else to the video element itself.

Reply:
I gotta say I was enthused about the possibility of simplifying with this symmetry, but I agree that for now we should only make `style` and `className` special.